### ISSUE TITLE
Insufficient system memory or apptainer/singularity platform

### ISSUE THREAD

**Issue by hosseinvk** [vdemichev/DiaNN#1236]
Hello Vadim,

Thank you for all the support around DIANN on GitHub.

I'm currently running DIANN 1.9.1 on our HPC and managed to reach the empirical library assembly step. A smaller test run with two samples completed successfully. However, when I attempted to process all (376) samples, the run crashed after about 20 minutes with an "insufficient system memory" error.

In my slurm script I have specified a large number of cpus and memories as follows:
```
#SBATCH --cpus-per-task=96
#SBATCH --mem=330G
```

These are shown in the log file as:
```
Logical CPU cores: 192
Thread number set to 96
```

Here is my code:
```
diann	--dir ${INPUT_DIR} \
		--lib ${spectral_lib} \
		--threads ${SLURM_CPUS_PER_TASK} \
		--out ${outdir}/${empirical_lib}.report \
		--out-lib ${outdir}/${empirical_lib} \
		--verbose 4 \
		--rt-profiling \
		--temp ${TEMP_DIR} \
		--use-quant \
		--gen-spec-lib \
		--mass-acc-ms1 ${ms1_acc} \
		${fasta_var_string} \
		--reannotate \
		${extra_flags} > ${DIANN_LOG}/3_assemble_empirical_lib.log 2>&1	
```

This is the log file:
```
INFO:    underlay of /etc/localtime required more than 50 (73) bind mounts
DIA-NN 1.9.1 (Data-Independent Acquisition by Neural Networks)
Compiled on Jul 15 2024 09:42:01
Current date and time: Thu Oct 31 11:57:35 2024
Logical CPU cores: 192
Thread number set to 96
The spectral library (if generated) will retain the original spectra but will include empirically-aligned RTs
Existing .quant files will be used
A spectral library will be generated
Library precursors will be reannotated using the FASTA database
WARNING: note the mass accuracy settings used by DIA-NN, automatic optimisation will not be performed as at least one of MS1/MS2 mass accuracies is user-provided
Mass accuracy will be fixed to 2e-05 (MS2) and 2e-05 (MS1)
WARNING: combining reuse of .quant files with automatic optimisation of mass accuracies or scan window will lead to results that are different from those of the original analysis that produced the .quant files and is therefore not recommended

376 files will be processed
[0:00] Loading spectral library /scratch2/pd35/hoss/p23_0611_e1/DIANN_Lib/Library_Kai.predicted.speclib
mimalloc: warning: mi_usable_size: pointer might not point to a valid heap region: 0x7f804c020000
(this may still be a valid very large allocation (over 64MiB))
mimalloc: warning: (yes, the previous pointer 0x7f804c020000 was valid after all)
mimalloc: warning: mi_usable_size: pointer might not point to a valid heap region: 0x7f804c020000
(this may still be a valid very large allocation (over 64MiB))
mimalloc: warning: (yes, the previous pointer 0x7f804c020000 was valid after all)
[0:05] Library annotated with sequence database(s): camprotR_240512_cRAP_20190401_full_tags.fasta; F:\Jsteele\Josie\peptide (1).fasta; F:\Jsteele\databases\JS_Human_SwissProt_Jun2023.fasta
[0:06] Spectral library loaded: 208224 protein isoforms, 217165 protein groups and 4957219 precursors in 1532933 elution groups.
[0:06] Loading FASTA /scratch2/pd35/hoss/p23_0611_e1/FASTA/iRT_peptides_fusion.fasta
[0:06] Loading FASTA /scratch2/pd35/hoss/p23_0611_e1/FASTA/JS_Human_SwissProt_Jun2023.fasta
[0:20] Loading FASTA /scratch2/pd35/hoss/p23_0611_e1/FASTA/Kai_iMetaLab_peptide.fasta
[0:23] Loading FASTA /scratch2/pd35/hoss/p23_0611_e1/FASTA/Universal_Contaminant_Protein_FASTA.fasta
[0:26] Reannotating library precursors with information from the FASTA database
[0:31] Finding proteotypic peptides (assuming that the list of UniProt ids provided for each peptide is complete)
[0:31] 4957219 precursors generated
[0:31] Gene names missing for some isoforms
[0:31] Library contains 207784 proteins, and 207441 genes
[0:33] Initialising library
[0:44] Cross-run analysis
[0:44] Reading quantification information: 376 files
[15:00] Quantifying peptides
mimalloc: warning: mi_usable_size: pointer might not point to a valid heap region: 0x7f7fd4020000
(this may still be a valid very large allocation (over 64MiB))
mimalloc: warning: (yes, the previous pointer 0x7f7fd4020000 was valid after all)
mimalloc: warning: mi_usable_size: pointer might not point to a valid heap region: 0x7f7fd4020000
(this may still be a valid very large allocation (over 64MiB))
mimalloc: warning: (yes, the previous pointer 0x7f7fd4020000 was valid after all)
mimalloc: warning: mi_usable_size: pointer might not point to a valid heap region: 0x7f7eec020000
(this may still be a valid very large allocation (over 64MiB))
mimalloc: warning: (yes, the previous pointer 0x7f7eec020000 was valid after all)
mimalloc: warning: mi_usable_size: pointer might not point to a valid heap region: 0x7f7eec020000
(this may still be a valid very large allocation (over 64MiB))
mimalloc: warning: (yes, the previous pointer 0x7f7eec020000 was valid after all)
mimalloc: warning: mi_free: pointer might not point to a valid heap region: 0x7f7fd4020000
(this may still be a valid very large allocation (over 64MiB))
mimalloc: warning: (yes, the previous pointer 0x7f7fd4020000 was valid after all)
mimalloc: warning: mi_usable_size: pointer might not point to a valid heap region: 0x7f7fd4020000
(this may still be a valid very large allocation (over 64MiB))
mimalloc: warning: (yes, the previous pointer 0x7f7fd4020000 was valid after all)
mimalloc: error: (mi_)aligned_alloc requires the size to be an integral multiple of the alignment (size 3395632136, alignment 16)
ERROR: insufficient system memory
ERROR: algorithmic failure: src/buffer.h: 89
/apps/diann/1.9.1/bin/diann: line 76: 3817474 Illegal instruction     (core dumped) apptainer exec "$sif_image" diann-linux $@
```

I did search for insufficient system memory issue on GitHub and it seems things were already resolved in `DIAN 1.9.1`.

My question is whether this error is truly related to memory size, or if it might indicate an issue with the DIANN module on our HPC, despite the successful run with two samples suggesting otherwise.

Many thanks for your help on this post.
Hossein

**Comment by vdemichev** [comment]
Hi Hossein,

Many thanks for reporting this :) 
You are right, it's indeed an error in specifically the Linux version of DIA-NN code. What happens, when I replaced the Windows function _aligned_malloc() with the Linux-compatible std::aligned_alloc(), I did not notice that the latter requires the requested memory amount to be divisible by the alignment, i.e. 16 in this case and will report an error if it's not. It's supposed to behave like that because it's mandated by the C++ standard for this function (but not on Windows). Will fix in the next release (soon).

Best,
Vadim
